File: src/lib/cache.ts
import { mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";

// Dependency-light persistent KV: one JSON file per entry under .cache/.
// No Redis/SQLite for a personal tool. Shared by CLI and the web server fns.
const CACHE_DIR = resolve(process.cwd(), ".cache");

interface Entry<T> {
  expiresAt: number; // epoch ms; 0 = never expires
  value: T;
}

function pathFor(namespace: string, key: string): string {
  const safe = key.replace(/[^a-zA-Z0-9._-]/g, "_");
  return resolve(CACHE_DIR, namespace, `${safe}.json`);
}

export function readCache<T>(namespace: string, key: string): T | null {
  try {
    const entry = JSON.parse(readFileSync(pathFor(namespace, key), "utf8")) as Entry<T>;
    if (entry.expiresAt && Date.now() > entry.expiresAt) return null;
    return entry.value;
  } catch {
    return null;
  }
}

export function writeCache<T>(namespace: string, key: string, value: T, ttlMs: number): void {
  try {
    const file = pathFor(namespace, key);
    mkdirSync(dirname(file), { recursive: true });
    const entry: Entry<T> = { expiresAt: ttlMs > 0 ? Date.now() + ttlMs : 0, value };
    writeFileSync(file, JSON.stringify(entry));
  } catch {
    // A cache write failure must never break the call it wraps.
  }
}

// Read-through cache. Returns the cached value (hit) or runs `produce`, stores
// it, and returns it (miss). `fresh` forces a miss to deliberately bypass.
export async function withCache<T>(
  namespace: string,
  key: string,
  ttlMs: number,
  produce: () => Promise<T>,
  opts?: { fresh?: boolean },
): Promise<{ value: T; hit: boolean }> {
  if (!opts?.fresh) {
    const cached = readCache<T>(namespace, key);
    if (cached !== null) return { value: cached, hit: true };
  }
  const value = await produce();
  writeCache(namespace, key, value, ttlMs);
  return { value, hit: false };
}

File: app/cache.py
"""Dependency-light persistent KV: one JSON file per entry under .cache/.

No Redis/SQLite for transient API/report caches — a personal tool wants zero
infra. Shared by the CLI and the web server. Port of the original lib/cache.ts.
"""

from __future__ import annotations

import asyncio
import json
import re
import time
from pathlib import Path
from typing import Any, Awaitable, Callable, TypeVar

from .config import CACHE_DIR

T = TypeVar("T")

_UNSAFE = re.compile(r"[^a-zA-Z0-9._-]")
_inflight: dict[str, tuple[asyncio.Lock, int]] = {}


def _path_for(namespace: str, key: str) -> Path:
    safe = _UNSAFE.sub("_", key)
    return CACHE_DIR / namespace / f"{safe}.json"


def read_cache(namespace: str, key: str) -> Any | None:
    try:
        entry = json.loads(_path_for(namespace, key).read_text())
    except (OSError, ValueError):
        return None
    expires_at = entry.get("expiresAt", 0)
    if expires_at and time.time() * 1000 > expires_at:
        return None
    return entry.get("value")


def write_cache(namespace: str, key: str, value: Any, ttl_ms: float) -> None:
    try:
        path = _path_for(namespace, key)
        path.parent.mkdir(parents=True, exist_ok=True)
        expires_at = time.time() * 1000 + ttl_ms if ttl_ms > 0 else 0
        path.write_text(json.dumps({"expiresAt": expires_at, "value": value}))
    except OSError:
        # A cache write failure must never break the call it wraps.
        pass


async def with_cache(
    namespace: str,
    key: str,
    ttl_ms: float,
    produce: Callable[[], Awaitable[T]],
    *,
    fresh: bool = False,
) -> tuple[T, bool]:
    """Read-through cache. Returns ``(value, hit)``. ``fresh`` forces a miss."""
    if not fresh:
        cached = read_cache(namespace, key)
        if cached is not None:
            return cached, True

    inflight_key = f"{namespace}:{key}"
    lock, refcount = _inflight.get(inflight_key, (asyncio.Lock(), 0))
    _inflight[inflight_key] = (lock, refcount + 1)
    try:
        async with lock:
            if not fresh:
                cached = read_cache(namespace, key)
                if cached is not None:
                    return cached, True
            value = await produce()
            if value is not None:
                write_cache(namespace, key, value, ttl_ms)
            return value, False
    finally:
        lock2, refcount2 = _inflight[inflight_key]
        if refcount2 <= 1:
            _inflight.pop(inflight_key, None)
        else:
            _inflight[inflight_key] = (lock2, refcount2 - 1)
